aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -586,5 +586,81 @@ Tensor &nuclear_norm_out(Tensor& result, const Tensor& self, bool keepdim) {
   return at::sum_out(result, std::get<1>(at::svd(self)), 0, keepdim);
 }
 
+Tensor _chain_matmul_general(TensorList matrices, std::vector<std::vector<int64_t>>& order, int64_t i, int64_t j) {
+  if (i == j)
+    return matrices[i];
+  else
+    return at::mm(_chain_matmul_general(matrices, order, i, order[i][j]), _chain_matmul_general(matrices, order, order[i][j] + 1, j));
+}
+
+// Why the separate implementation for 3 matrices?
+// The logic for three matrices is much faster when done directly
+// Requires 1 comparison to 4 comparisons and lesser arithmetic operations
+Tensor _chain_matmul_three_matrices(TensorList matrices) {
+  int64_t a = matrices[0].size(0);  // This is the first dimension
+  int64_t b = matrices[1].size(0);  // This is the common dimension between the first two matrices
+  int64_t c = matrices[2].size(0);  // This is the common dimension between the last two matrices
+  int64_t d = matrices[2].size(1);  // This is the last dimension
+
+  // The matrices are of size (a x b), (b x c), (c x d)
+  // cost_1 is the cost of parenthesizing (a x b) and (b x c) and then combining (c x d)
+  // cost_2 is the cost of parenthesizing (b x c) and (c x d) and then combining (a x b)
+  int64_t cost_1 = (a * c) * (b + d);
+  int64_t cost_2 = (b * d) * (a + c);
+
+  if (cost_1 > cost_2) {
+    return at::mm(matrices[0], at::mm(matrices[1], matrices[2]));
+  } else {
+    return at::mm(at::mm(matrices[0], matrices[1]), matrices[2]);
+  }
+}
+
+Tensor chain_matmul(TensorList matrices) {
+  AT_CHECK(matrices.size() >= 2, "Expecting at least 2 matrices");
+  checkAllSameDim(matrices, 2);
+
+  if (matrices.size() == 2) {
+    return at::mm(matrices[0], matrices[1]);
+  } else if (matrices.size() == 3) {
+    return _chain_matmul_three_matrices(matrices);
+  } else {
+
+    // Following the algorithm in Chapter 15.2 : Introduction to Algorithms, Cormen et al.
+    // Minor modifications have be made to accommodate zero-indexing
+    auto n = matrices.size();
+    std::vector<int64_t> p;
+    p.push_back(matrices[0].size(0));
+    for (int64_t i = 0; i < n; i++) {
+      p.push_back(matrices[i].size(1));
+    }
+
+    // Cost matrix
+    std::vector<std::vector<double>> m(n, std::vector<double>(n, 0));
```

Comment:
Why is this a `double` instead of `int64_t`? Double arithmetic is a lot more expensive.

Reply:
Also some brief explanation on what `m` are `p` store would be nice so people won't have to find the book that isn't freely distributed. :)